modules/openapi-generator/src/main/resources/python/model.mustache

```diff
@@ -33,48 +33,104 @@ class {{classname}}(object):
+
+{{#optionalVars}}{{^hasMore}}        Keyword Args:{{/hasMore}}{{/optionalVars}}{{#optionalVars}}
+            {{name}} ({{dataType}}):{{#description}} {{description}}.{{/description}} [optional]{{#defaultValue}} if omitted the server will use the default value of {{{defaultValue}}}{{/defaultValue}}{{/optionalVars}}
+
+        """  # noqa: E501
+
+        self._data_store = {}
+
         self.discriminator = {{#discriminator}}'{{{discriminatorName}}}'{{/discriminator}}{{^discriminator}}None{{/discriminator}}
-{{#vars}}{{#-first}}
-{{/-first}}
-{{#required}}
-        self.{{name}} = {{name}}
-{{/required}}
-{{^required}}
-{{#isNullable}}
-        self.{{name}} = {{name}}
-{{/isNullable}}
-{{^isNullable}}
-        if {{name}} is not None:
-            self.{{name}} = {{name}}
-{{/isNullable}}
-{{/required}}
-{{/vars}}
+{{#requiredVars}}
+        self.__setitem__('{{name}}', {{name}})
+{{/requiredVars}}
+
+        for var_name, var_value in six.iteritems(kwargs):
+            self.__setitem__(var_name, var_value)
+
+    def recursive_type(self, item):
+        """Gets a string describing the full the recursive type of a value"""
+        item_type = type(item)
+        if item_type == dict:
+            child_key_types = set()
+            child_value_types = set()
+            for child_key, child_value in six.iteritems(item):
+                child_key_types.add(self.recursive_type(child_key))
+                child_value_types.add(self.recursive_type(child_value))
+            if child_key_types != set(['str']):
+                raise ValueError('Invalid dict key type. All Openapi dict keys must be strings')
+            child_value_types = '|'.join(sorted(list(child_value_types)))
+            return "dict(str, {0})".format(child_value_types)
+        elif item_type == list:
+            child_value_types = set()
+            for child_item in item:
+                child_value_types.add(self.recursive_type(child_item))
+            child_value_types = '|'.join(sorted(list(child_value_types)))
+            return "list[{0}]".format(child_value_types)
+        else:
+            return type(item).__name__
+
+    def __setitem__(self, name, value):
+        check_type = False
+        if name in self.openapi_types:
+            required_type = self.openapi_types[name]
+        else:
+{{#additionalProperties}}
+            check_type = True
+            required_type = self.additional_properties_type
+{{/additionalProperties}}
+{{^additionalProperties}}
+            raise KeyError("{0} has no key '{1}'".format(
+                type(self).__name__, name))
+{{/additionalProperties}}
+
+        passed_type = self.recursive_type(value)
+        if type(name) != str:
+            raise TypeError('Variable name must be type string and %s was not' % name)
+        elif passed_type != required_type and check_type:
```

Comment:
Thank you very much for creating this pull request! I'm the author of https://github.com/OpenAPITools/openapi-generator/issues/2028 and https://github.com/OpenAPITools/openapi-generator/pull/2029 and I think the solution you have here is cleaner and more robust. 

Just one minor issue I've noticed so far. I tested it with my use case of
```javascript
additionalProperties:
            type: "object"
```
which caused this error to be raised when doing `model["foo"] = "bar"`with `ValueError: Variable value must be type object but you passed in str`. It looks like the type checking isn't taking into account subclasses. I would really appreciate if you could address this use case. I would be happy to test for you as well.

Reply:
Hi Ivan. You're welcome. Thank you for yours. We both had the same idea to add a dict to store the values, this pull request just implemented it as a general case. Thanks for the heads up about that failing case. To fix it, I will add more samples. For the example you gave the code was right to raise an error, because you should have passed in:
model["foo"] = {"some_var": "bar"}
But the error message was wrong in the type that it was demanding. It should have wanted a dict rather than an object. We collapse complex schema types down elsewhere, so I will find that code and use it.

Reply:
The intent of additionalProperties with type object is that any value, regardless of type, is accepted. Upon review of the OpenAPI spec I should be using `additionalProperties: true` to represent that intent (source: https://swagger.io/docs/specification/data-models/dictionaries/). I will try this case out for us.

Reply:
When `additionalProperties: true` is used, the codegen treats it as if `additionalProperties` is not enabled. I suspect that the fact that its type isn't defined causes it to be ignored.

Reply:
Ivan, what does inspecting the values in the openapi spec show when we use:
- additionalProperties: true
- additionalProperties: {}
If we see it in those results, we can handle it in the java code.
One can inspect the values in the openapi spec with the below line:
```
java -DdebugOpenAPI -jar modules/openapi-generator-cli/target/openapi-generator-cli.jar generate -t modules/openapi-generator/src/main/resources/python -i modules/openapi-generator/src/test/resources/2_0/petstore-with-fake-endpoints-models-for-testing.yaml -g python -o samples/client/petstore/python -DpackageName=petstore_api
```

Reply:
Modifying `additionalProperties` in `AdditionalPropertiesClass` in `modules/openapi-generator/src/test/resources/2_0/petstore-with-fake-endpoints-models-for-testing.yaml` to `additionalProperties: {}` results in 
```javascript
      "AdditionalPropertiesClass" : {
        "type" : "object",
        "properties" : {
          "map_property" : {
            "type" : "object",
            "additionalProperties" : {
              "type" : "string"
            }
          },
          "map_of_map_property" : {
            "type" : "object",
            "additionalProperties" : {
              "type" : "object",
              "additionalProperties" : {
                "type" : "string"
              }
            }
          }
        },
        "additionalProperties" : {
          "type" : "object"
        }
```
Note `"type" : "object"`.

Using `additionalProperties: true` results in
```javascript
      "AdditionalPropertiesClass" : {
        "type" : "object",
        "properties" : {
          "map_property" : {
            "type" : "object",
            "additionalProperties" : {
              "type" : "string"
            }
          },
          "map_of_map_property" : {
            "type" : "object",
            "additionalProperties" : {
              "type" : "object",
              "additionalProperties" : {
                "type" : "string"
              }
            }
          }
        }
      }
```
Note missing `additionalProperties`.

For my specific case, I am fine with using `additionalProperties: {}`, but that results in the same problematic behavior above where `object` is being treated as expected `dict`.

Reply:
Ivan I am figuring out how to add this empty object -> all types.
This is also complicated by the fact that any type can have nullable set to true, which would mean that we could accept types like (string) (string OR None) or all the way up to (string, boolean, integer, float, list, dict, None).

For now I will leave nullable out because it is a 3.0 spec feature.
We still will have cases were "str|bool|int|float|list|dict" would be the type or you could have that nested as the innermost value in a nested list or dict schema, like:
dict(str, dict(str, str|bool|int|float|list|dict))

Reply:
@wing328 @ivan-gomes I am still working on this. Please refrain from merging until I have updated this PR. My update will include the type=object additionalProperties feature.

Reply:
Roger. Please let me know if there is anything I can do to help.

Reply:
@ivan-gomes will do.
Still to do is:
- Move model utils to the model `__init__` file
  - recursive_type
  - valid_type
  - get_types_remainder
  - to_dict
- Add additionalproperties map dict so variable names can be transformed for serialization/deserialization
- Ensure serialization works w/ tests
- Ensure deserialization works w/ tests
- Have serialization use model.to_dict

Reply:
@wing328 @ivan-gomes I am nearing completion on this task.
To get this feature to work I added typing of all parameters and discriminators using real python classes. Soon I will get this PR submitted.